backends/platform/android/org/scummvm/scummvm/SAFFSTree.java

```diff
@@ -0,0 +1,439 @@
+					ourFlags |= SAFFSNode.WRITABLE;
+				}
+				if ((flags & DocumentsContract.Document.FLAG_VIRTUAL_DOCUMENT) == 0) {
+					ourFlags |= SAFFSNode.READABLE;
+				}
+
+				SAFFSNode newnode = new SAFFSNode(node, node._path + "/" + displayName, documentId, ourFlags);
+				_cache.put(newnode._path, newnode);
+				return newnode;
+			}
+		} catch (Exception e) {
+			Log.w(ScummVM.LOG_TAG, "Failed query: " + e);
+		} finally {
+			if (c != null) {
+				c.close();
+			}
+		}
+
+		_cache.put(node._path + "/" + name, NOT_FOUND_NODE);
+		return null;
+	}
+
+	public SAFFSNode createDirectory(SAFFSNode node, String name) {
+		final ContentResolver resolver = _context.getContentResolver();
+		final Uri parentUri = DocumentsContract.buildDocumentUriUsingTree(_treeUri, node._documentId);
+		Uri newDirUri;
+
+		try {
+			newDirUri = DocumentsContract.createDocument(resolver, parentUri, DocumentsContract.Document.MIME_TYPE_DIR, name);
+		} catch(FileNotFoundException e) {
+			return null;
+		}
+		if (newDirUri == null) {
+			return null;
+		}
+
+		final String documentId = DocumentsContract.getDocumentId(newDirUri);
+
+		final SAFFSNode newnode = new SAFFSNode(node, node._path + "/" + name, documentId, 0);
+		// Update flags
+		final String realName = stat(_root);
+		if (realName == null) {
+			return null;
+		}
+		// Unlikely but...
+		if (!realName.equals(name)) {
+			newnode._path = node._path + "/" + realName;
+		}
+
+		_cache.put(newnode._path, newnode);
+
+		return newnode;
+	}
+
+	public SAFFSNode createFile(SAFFSNode node, String name) {
+		final ContentResolver resolver = _context.getContentResolver();
+		final Uri parentUri = DocumentsContract.buildDocumentUriUsingTree(_treeUri, node._documentId);
+		Uri newFileUri;
+
+		try {
+			newFileUri = DocumentsContract.createDocument(resolver, parentUri, "application/octet-stream", name);
+		} catch(FileNotFoundException e) {
+			return null;
+		}
+		if (newFileUri == null) {
+			return null;
+		}
+
+		String documentId = DocumentsContract.getDocumentId(newFileUri);
+
+		SAFFSNode newnode = new SAFFSNode(node, node._path + "/" + name, documentId, 0);
+		// Update flags
+		String realName = stat(_root);
+		if (realName == null) {
+			return null;
+		}
+		// Unlikely but...
+		if (!realName.equals(name)) {
+			newnode._path = node._path + "/" + realName;
+		}
+
+		_cache.put(newnode._path, newnode);
+
+		return newnode;
+	}
+
+	public int createReadStream(SAFFSNode node) {
+		final ContentResolver resolver = _context.getContentResolver();
+		final Uri uri = DocumentsContract.buildDocumentUriUsingTree(_treeUri, node._documentId);
+
+		ParcelFileDescriptor pfd;
+		try {
+			pfd = resolver.openFileDescriptor(uri, "r");
+		} catch(FileNotFoundException e) {
+			return -1;
+		}
+		if (pfd == null) {
+			return -1;
+		}
+
+		return pfd.detachFd();
+	}
+
+	public int createWriteStream(SAFFSNode node) {
```

Comment:
Since this code is 99% the same as `createReadStream()` above, perhaps it would make sense to reuse it using an internal function?

Reply:
Indeed. Done.